### PR TITLE
Allow warm-start in diffcp

### PR DESCRIPTION
**Pull request by bstellato**
This PR allows to warm-start SCS in diffcp. Note that the `warm_start` option in diffcp now conflicts with the standard cvxpy `warm_start` option which is True/False.

Also, it would be good to check the type of the `warm_start` entry in diffcp to avoid errors:
https://github.com/cvxgrp/diffcp/blob/2e0d3f7337e19509bd312272ed59be1f93104ff2/diffcp/cone_program.py#L248-L251

### CODE REVIEW

File: cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py [hunk 1]
```diff
@@ -77,8 +77,13 @@ def solve_via_data(self, data, warm_start, verbose, solver_opts,
         cones = scs_conif.dims_to_solver_dict(data[ConicSolver.DIMS])
         # Default to eps = 1e-4 instead of 1e-3.
         solver_opts['eps'] = solver_opts.get('eps', 1e-4)
+        # Ignore True/False warm_start. Warm-start only with values
```
Comment (akshayka) [inline cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py]: I don't think we should ignore True/False warm_start. We should make warm-starting diffcp (through CVXPY) identical to warm-starting SCS: https://sourcegraph.com/github.com/cvxgrp/cvxpy@852951e19c5a42d88d65a45d4ef69610d0ca24c9/-/blob/cvxpy/reductions/solvers/conic_solvers/scs_conif.py#L289-293

Comment (bstellato) [reply cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py]: That sounds much better. I have edited the PR accordingly.

**Comment by akshayka** [comment]
Looks great. One more thing (& probably the last thing)  --- can you add a unit test that exercises the warm start code path? Something simple is fine, e.g., similar to https://sourcegraph.com/github.com/cvxgrp/cvxpy@af2f1e8eddb2997ac52d105f6e81d5d63582180d/-/blob/cvxpy/tests/test_scs.py#L186-196 (without the print statement).

The test could go in test_scs.py or test_derivative.py.

Also, flake8:
./cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py:83:16: E111 indentation is not a multiple of four

File: cvxpy/tests/test_scs.py [hunk 2]
```diff
@@ -196,6 +196,19 @@ def test_warm_start(self):
...
+        """
+        x = cp.Variable(10)
+        obj = cp.Minimize(cp.sum(cp.exp(x)))
+        prob = cp.Problem(obj, [cp.sum(x) == 1])
+        result = prob.solve(solver=cp.DIFFCP, eps=1e-4)
+        time = prob.solver_stats.solve_time
+        result2 = prob.solve(solver=cp.DIFFCP, warm_start=True, eps=1e-4)
+        time2 = prob.solver_stats.solve_time
+        self.assertAlmostEqual(result2, result, places=2)
+        print(time > time2)
```
Comment (akshayka) [inline cvxpy/tests/test_scs.py]: Delete the print statement, please.

File: cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py [hunk 3]
```diff
@@ -77,8 +77,16 @@ def solve_via_data(self, data, warm_start, verbose, solver_opts,
         cones = scs_conif.dims_to_solver_dict(data[ConicSolver.DIMS])
         # Default to eps = 1e-4 instead of 1e-3.
         solver_opts['eps'] = solver_opts.get('eps', 1e-4)
+        warm_start_dict = None
```
Comment (akshayka) [inline cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py]: warm_start_dict -> warm_start_tuple

File: cvxpy/tests/test_scs.py [hunk 4]
```diff
@@ -196,6 +196,19 @@ def test_warm_start(self):
...
         print(time > time2)
 
+    def test_warm_start_diffcp(self):
+        """Test warm starting in diffcp.
+        """
+        x = cp.Variable(10)
+        obj = cp.Minimize(cp.sum(cp.exp(x)))
+        prob = cp.Problem(obj, [cp.sum(x) == 1])
+        result = prob.solve(solver=cp.DIFFCP, eps=1e-4)
+        time = prob.solver_stats.solve_time
```
Comment (akshayka) [inline cvxpy/tests/test_scs.py]: Delete this line

File: cvxpy/tests/test_scs.py [hunk 5]
```diff
@@ -196,6 +196,19 @@ def test_warm_start(self):
...
+        result2 = prob.solve(solver=cp.DIFFCP, warm_start=True, eps=1e-4)
+        time2 = prob.solver_stats.solve_time
```
Comment (akshayka) [inline cvxpy/tests/test_scs.py]: Delete this line

**Comment by bstellato** [comment]
Should be done now. Let me know if there is anything else to change. Thanks!

**Pull request merged**